### PR TITLE
Restore page

### PR DESCRIPTION
**Pull request by song50119**
### Add restore page

Issue #16

### What's new

Created restore page

Screen Shot 2020-02-05 at 23 43 56

### CODE REVIEW

**Comment by song50119** [comment]
Add radio buttons page "Spending password and PIN"
Screen Shot 2020-02-07 at 21 47 58

**Review by oluwandabira — DISMISSED** [review]

File: ui/page/landing.go [hunk 1]
```diff
@@ -75,12 +75,18 @@ func (pg *Landing) Draw(gtx *layout.Context) (res interface{}) {
...
+					Next:    RestoreID,
+				}
 			}
 			pg.restoreBtn.Layout(gtx, pg.restoreWdg)
+			if pg.restoreWdg.Clicked(gtx) {
+				res = EventNav{
+					Current: LandingID,
+					Next:    RestoreID,
+				}
+			}
```
Comment (oluwandabira) [inline ui/page/landing.go]: `if pg.restoreWdg.Clicked(gtx)` is duplicated

File: ui/page/restore.go [hunk 2]
```diff
@@ -0,0 +1,446 @@
...
+func (pg *modal) initModal(theme *materialplus.Theme) {
+	pg.confirmErrorMsg = theme.Label(unit.Dp(16), "")
+	pg.confirmErrorMsg.Color = DangerColor
+
+	pg.inputPassword = theme.Editor("Password...")
+	pg.editorPassword = &widget.Editor{SingleLine: true}
+	pg.inputConfirmPassword = theme.Editor("Confirm password...")
+	pg.editorConfirmPassword = &widget.Editor{SingleLine: true}
+	pg.inputSendingPin = theme.Editor("Input PIN...")
+	pg.editorSendingPin = &widget.Editor{SingleLine: true}
```
Comment (oluwandabira) [inline ui/page/restore.go]: No more pin

File: ui/page/restore.go [hunk 3]
```diff
@@ -0,0 +1,446 @@
...
+}
+
+func (pg *modal) validatePassword() bool {
+	pg.confirmErrorMsg.Text = ""
+	if pg.editorPassword.Text() == "" {
+		pg.confirmErrorMsg.Text = "Please enter password"
+		return false
+	}
+	if pg.editorConfirmPassword.Text() == "" {
+		pg.confirmErrorMsg.Text = "Please enter confirm password"
```
Comment (oluwandabira) [inline ui/page/restore.go]: "Please enter confirm password" => "Confirm password"

File: ui/page/restore.go [hunk 4]
```diff
@@ -0,0 +1,446 @@
...
+		return false
+	}
+	if pg.editorConfirmPassword.Text() != pg.editorPassword.Text() {
+		pg.confirmErrorMsg.Text = "Password and confirm password do not match"
```
Comment (oluwandabira) [inline ui/page/restore.go]: "Password and confirm password do not match" => "Passwords do no match"

File: ui/page/restore.go [hunk 5]
```diff
@@ -0,0 +1,446 @@
...
+
+		pg.isRestoring = true
+		pg.submitBtn.Text = "Please wait..."
+		pg.submitBtn.Background = pg.theme.Color.Hint
+		pg.wal.RestoreWallet(pg.wordsFromEditors(), pg.editorSendingPin.Text())
+
+		go func() {
+			resp := <-pg.wal.Send
+
+			if resp.Err != nil {
```
Comment (oluwandabira) [inline ui/page/restore.go]: See #32 for how to handle non-blocking wallet commands. TLDR: page issues the command, window receives the result.

**Review by oluwandabira — CHANGES_REQUESTED** [review]

File: ui/page/restore.go [hunk 6]
```diff
@@ -0,0 +1,470 @@
...
+			}
+		case wallet.InternalWalletError:
+			pg.states[StateWalletRespondedError] = nil
+		default:
+			pg.confirmErrorMsg.Text = "Restore failed, please try later"
+		}
+
+		pg.submitBtn.Text = "Restore"
+		pg.submitBtn.Background = pg.theme.Color.Primary
+		pg.states[StateWalletRespondedError] = nil
```
Comment (oluwandabira) [inline ui/page/restore.go]: delete the key

File: ui/page/restore.go [hunk 7]
```diff
@@ -0,0 +1,470 @@
...
+	}
+
+	response := pg.states[StateWalletRestored].(*wallet.Restored)
+
+	if response != nil {
+		pg.cleanAllInputs()
+		pg.submitBtn.Text = "Restore"
+		pg.submitBtn.Background = pg.theme.Color.Primary
+		pg.showModal = false
+		pg.states[StateWalletRestored] = nil
```
Comment (oluwandabira) [inline ui/page/restore.go]: delete the key

File: ui/page/restore.go [hunk 8]
```diff
@@ -0,0 +1,470 @@
...
+				if i < 32 {
+					pg.editors[i+1].Focus()
+				}
+			}
+		}
+	}
+}
+
+func (pg *Restore) onRestored() {
+	if pg.states[StateWalletRestored] == nil {
```
Comment (oluwandabira) [inline ui/page/restore.go]: check for the existence of the key

File: ui/page/restore.go [hunk 9]
```diff
@@ -0,0 +1,470 @@
...
+		pg.submitBtn.Text = "Restore"
+		pg.submitBtn.Background = pg.theme.Color.Primary
+		pg.showModal = false
+		pg.states[StateWalletRestored] = nil
+		pg.isRestoring = false
+	}
+}
+
+func (pg *Restore) onRestoredError() {
+	if pg.states[StateWalletRespondedError] == nil {
```
Comment (oluwandabira) [inline ui/page/restore.go]: check for the existence of the key

**Review by oluwandabira — CHANGES_REQUESTED** [review]

**Review by oluwandabira — DISMISSED** [review]

File: ui/restore_page.go [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+						}),
+						layout.Rigid(func() {
+							inputsGroup(win, inputGroupContainerRight, 17, 16)
+						}),
+					)
+				})
+			}),
+			layout.Rigid(func() {
+				layout.Center.Layout(win.gtx, func() {
+					win.gtx.Constraints.Width.Min = 210
```
Comment (oshorefueled) [inline ui/restore_page.go]: you should avoid using hardcoded values in your code.

**Review by oshorefueled — CHANGES_REQUESTED** [review]

File: ui/window.go [hunk 11]
```diff
@@ -113,7 +114,9 @@ func (win *Window) Loop(shutdown chan int) {
 
 				win.HandleInputs()
 				evt.Frame(win.gtx.Ops)
-
+			case key.Event:
+				win.combined.keyEvent = &evt
+				win.KeysEventsHandler()
```
Comment (oluwandabira) [inline ui/window.go]: Pass the key event as an argument to `win.KeysEventsHandler`

File: ui/handle_input.go [hunk 12]
```diff
@@ -155,3 +174,78 @@ func (win *Window) resetPasswords() {
...
+			if win.combined.seedEditorsHandlerIndex < len(win.inputs.seeds)-1 {
+				win.inputs.seeds[win.combined.seedEditorsHandlerIndex+1].Focus()
+			}
+		}
+	}
+}
+
+// KeysEventsHandler handlers all key events when typing editor, if Tab pressed will putting first word
+// from the list of suggestion to the editor
+func (win *Window) KeysEventsHandler() {
```
Comment (oluwandabira) [inline ui/handle_input.go]: This should take the key event as an argument since it is only used within this scope.

File: ui/widgets.go [hunk 13]
```diff
@@ -1,39 +1,51 @@
...
+
+	seeds []widget.Editor
 }
 
 type combined struct {
 	sel *decredmaterial.Select
+
+	keyEvent                *key.Event
+	seedEditorsHandlerIndex int
+	seedsSuggestions        []string
```
Comment (oluwandabira) [inline ui/widgets.go]: Not widgets.

File: ui/widgets.go [hunk 14]
```diff
@@ -1,39 +1,51 @@
...
 
 	spendingPassword, matchSpending, renameWallet widget.Editor
+
+	seeds []widget.Editor
 }
 
 type combined struct {
 	sel *decredmaterial.Select
+
+	keyEvent                *key.Event
```
Comment (oluwandabira) [inline ui/widgets.go]: Unnecessary.

File: ui/widgets.go [hunk 15]
```diff
@@ -1,39 +1,51 @@
...
+	seedEditorsHandlerIndex int
+	seedsSuggestions        []string
+	seedsSuggestionsBtn     [2]widget.Button
```
Comment (oluwandabira) [inline ui/widgets.go]: Input widget.

**Review by oluwandabira — CHANGES_REQUESTED** [review]

**Review by oluwandabira — DISMISSED** [review]

**Review by metaclips — APPROVED** [review]
UI functionality looks good

**Review by oluwandabira — APPROVED** [review]

**Review by oshorefueled — APPROVED** [review]

**Pull request merged**